modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IgniteIndexing.java

```diff
@@ -0,0 +1,329 @@
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void store(@Nullable String cacheName, Object key, Object val,
+        long expirationTime) throws IgniteSpiException {
+        throw new IgniteSpiException("Not implemented.");
+    }
+
+    /** {@inheritDoc} */
+    @Override public void store(GridCacheContext<?, ?> cctx, CacheDataRow newRow, @Nullable CacheDataRow prevRow,
+        boolean prevRowAvailable)
+        throws IgniteSpiException {
+        try {
+            updateIndexes(cctx.name(), newRow, prevRow, prevRowAvailable);
+
+        } catch (IgniteCheckedException e) {
+            throw new IgniteSpiException("Failed to store row in cache", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void store(Collection<? extends Index> idxs, CacheDataRow newRow, @Nullable CacheDataRow prevRow,
+        boolean prevRowAvailable) throws IgniteSpiException {
+        IgniteCheckedException err = null;
+
+        ddlLock.readLock().lock();
+
+        try {
+            for (Index idx : idxs)
+                err = updateIndex(idx, newRow, prevRow, prevRowAvailable, err);
+
+            if (err != null)
+                throw err;
+
+        } catch (IgniteCheckedException e) {
+            throw new IgniteSpiException("Failed to store row in index", e);
+
+        } finally {
+            ddlLock.readLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void remove(@Nullable String cacheName, Object key) throws IgniteSpiException {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public void remove(String cacheName, @Nullable CacheDataRow prevRow) throws IgniteSpiException {
+        try {
+            updateIndexes(cacheName, null, prevRow, true);
+
+        } catch (IgniteCheckedException e) {
+            throw new IgniteSpiException("Failed to remove row in cache", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public Index createIndex(GridCacheContext<?, ?> cctx, IndexFactory factory, IndexDefinition definition) {
+        ddlLock.writeLock().lock();
+
+        try {
+            String cacheName = definition.getIdxName().cacheName();
+
+            cacheToIdx.putIfAbsent(cacheName, new ConcurrentHashMap<>());
+
+            String uniqIdxName = definition.getIdxName().fullName();
+
+            // GridQueryProcessor already checked schema operation for index duplication.
+            assert cacheToIdx.get(cacheName).get(uniqIdxName) == null : "Duplicated index name " + uniqIdxName;
+
+            Index idx = factory.createIndex(cctx, definition);
+
+            cacheToIdx.get(cacheName).put(uniqIdxName, idx);
+
+            idxDefs.put(idx.id(), definition);
+
+            return idx;
+
+        } finally {
+            ddlLock.writeLock().unlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void removeIndex(GridCacheContext<?, ?> cctx, IndexDefinition def, boolean softDelete) {
```

Comment:
`IndexName` should uniquely describe index, so let's use it instead of `IndexDefinition`